src/wp-includes/abilities-api/class-wp-abilities-category-registry.php

```diff
@@ -0,0 +1,247 @@
+			_doing_it_wrong(
+				__METHOD__,
+				/* translators: %s: Ability category slug. */
+				sprintf( esc_html__( 'Ability category "%s" not found.' ), esc_attr( $slug ) ),
+				'6.9.0'
+			);
+			return null;
+		}
+
+		$unregistered_category = $this->registered_categories[ $slug ];
+		unset( $this->registered_categories[ $slug ] );
+
+		return $unregistered_category;
+	}
+
+	/**
+	 * Retrieves the list of all registered categories.
+	 *
+	 * Do not use this method directly. Instead, use the `wp_get_ability_categories()` function.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @see wp_get_ability_categories()
+	 *
+	 * @return array<string,\WP_Ability_Category> The array of registered categories.
+	 */
+	public function get_all_registered(): array {
+		return $this->registered_categories;
+	}
+
+	/**
+	 * Checks if a category is registered.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @param string $slug The slug of the category.
+	 * @return bool True if the category is registered, false otherwise.
+	 */
+	public function is_registered( string $slug ): bool {
+		return isset( $this->registered_categories[ $slug ] );
+	}
+
+	/**
+	 * Retrieves a registered category.
+	 *
+	 * Do not use this method directly. Instead, use the `wp_get_ability_category()` function.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @see wp_get_ability_category()
+	 *
+	 * @param string $slug The slug of the registered category.
+	 * @return ?\WP_Ability_Category The registered category instance, or null if it is not registered.
+	 */
+	public function get_registered( string $slug ): ?WP_Ability_Category {
+		if ( ! $this->is_registered( $slug ) ) {
+			_doing_it_wrong(
+				__METHOD__,
+				/* translators: %s: Ability category slug. */
+				sprintf( esc_html__( 'Ability category "%s" not found.' ), esc_attr( $slug ) ),
+				'6.9.0'
+			);
+			return null;
+		}
+		return $this->registered_categories[ $slug ];
+	}
+
+	/**
+	 * Utility method to retrieve the main instance of the registry class.
+	 *
+	 * The instance will be created if it does not exist yet.
+	 *
+	 * @since 6.9.0
+	 *
+	 * @return \WP_Abilities_Category_Registry The main registry instance.
+	 */
+	public static function get_instance(): self {
+		if ( null === self::$instance ) {
+			self::$instance = new self();
+
+			/**
+			 * Fires when preparing ability categories registry.
+			 *
+			 * Categories should be registered on this action to ensure they're available when needed.
+			 *
+			 * @since 6.9.0
+			 *
+			 * @param \WP_Abilities_Category_Registry $instance Categories registry object.
+			 */
+			do_action( 'abilities_api_categories_init', self::$instance );
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Wakeup magic method.
+	 *
+	 * @since 6.9.0
+	 * @throws \LogicException If the registry is unserialized. This is a security hardening measure to prevent unserialization of the registry.
+	 */
+	public function __wakeup(): void {
+		throw new \LogicException( self::class . ' must not be unserialized.' );
```

Comment:
```suggestion
		throw new \LogicException( __CLASS__ . ' should never be unserialized.' );
```

nit: In other core code, we use `__CLASS__` with this phrase. This makes it consistent. 

See: https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/class-wp-block-bindings-source.php#L107 and https://github.com/WordPress/wordpress-develop/blob/trunk/src/wp-includes/html-api/class-wp-html-open-elements.php#L850 for two instances